vault/src/process.rs

```diff
@@ -0,0 +1,230 @@
+use std::{
+    fs::{self, File},
+    io::Write,
+    path::PathBuf,
+    process,
+    str::FromStr,
+};
+
+use runtime::AccountId;
+use service::Error;
+use sysinfo::{Pid, PidExt, ProcessExt, System, SystemExt};
+
+pub trait SystemProcess {
+    fn refresh_process(&mut self, pid: Pid) -> bool;
+    fn process_name(&self, pid: Pid) -> Result<String, Error>;
+}
+
+impl SystemProcess for System {
+    fn refresh_process(&mut self, pid: Pid) -> bool {
+        <Self as SystemExt>::refresh_process(self, pid)
+    }
+
+    fn process_name(&self, pid: Pid) -> Result<String, Error> {
+        let process = <Self as SystemExt>::process(self, pid).ok_or_else(|| Error::ProcessNotFound(pid.to_string()))?;
+        Ok(process.name().to_string())
+    }
+}
+
+#[derive(Debug)]
+pub struct PidFile {
+    path: PathBuf,
+}
+
+impl PidFile {
+    pub fn new(spec_name: &String, account_id: &AccountId) -> Self {
+        Self {
+            path: Self::compute_path(spec_name, account_id),
+        }
+    }
+
+    pub fn create(spec_name: &String, account_id: &AccountId, sys: &mut impl SystemProcess) -> Result<Self, Error> {
+        let pid_file = Self::new(spec_name, account_id);
+        if pid_file.path.exists() {
+            let pid = pid_file.pid()?;
+            if sys.refresh_process(pid) {
+                match pid_name_matches_existing_client(sys, pid) {
+                    Ok(true) => return Err(Error::ServiceAlreadyRunning(pid.to_string())),
+                    Ok(false) => (),
+                    // There is a very small chance that a `pid` with a running process is killed exactly before
+                    // `pid_name_matches_client` is run and `ProcessNotFound` is thrown. This is as if
+                    // the `refresh_process` check had returned `false` (the pidfile process not running)
+                    // so the error should not be propagated.
+                    Err(Error::ProcessNotFound(_)) => (),
+                    Err(e) => return Err(e),
+                }
+            }
+            tracing::warn!(
+                "Found existing PID file at: {}, but the process is no longer running.",
+                pid_file
+                    .path
+                    .clone()
+                    .into_os_string()
+                    .into_string()
+                    .map_err(|_| Error::OsStringError)?,
+            );
+        }
+
+        tracing::info!(
+            "Creating PID file at: {}",
+            pid_file
+                .path
+                .clone()
+                .into_os_string()
+                .into_string()
+                .map_err(|_| Error::OsStringError)?,
+        );
+        pid_file.set_pid(process::id())?;
+        Ok(pid_file)
+    }
+
+    pub fn compute_path(spec_name: &String, account_id: &AccountId) -> PathBuf {
+        let file_name = format!("{}_{}.pid", spec_name, account_id);
+        // Use the default temporary directory of the OS
+        std::env::temp_dir().join(file_name)
```

Comment:
I think it would make more sense to push this to `/var/run`. See: https://www.pathname.com/fhs/2.2/fhs-5.13.html

Reply:
However, `bitcoind` (for example on testnet) creates the file here: `~/.bitcoin/testnet3/bitcoind.pid`.

Reply:
`/var/run` requires sudo. `/run` [too](https://refspecs.linuxfoundation.org/FHS_3.0/fhs/ch03s15.html), which is what bitcoin-core [uses](https://github.com/bitcoin/bitcoin/blob/2557429d2b82fe5ba308b940bcf14c3047634345/contrib/init/bitcoind.service#L22), because that's the convention in the Filesystem Hierarchy Standard. 

The `/tmp` directory does not require sudo and its files are removed on reboot (in most distros). If we manage our own directory (e.g. `~/.$SPECNAME`) we wouldn't be able to guarantee that leftover pidfiles (those not cleaned up because the vault is killed with SIGKILL) are removed.

If we want the vault to require root permissions then we can go for `/run`, otherwise I think `/tmp` is the better option.

Reply:
It is also possible to use `/var/run/user/[$uid]` without root permissions I think, otherwise I agree that `/tmp` is a good alternative - though some distros do not always clean this upon boot. I believe it is also common practice to store this file in the user's home directory but as you pointed out that will not be removed automatically by the OS.

Reply:
I will go for `/tmp` in this PR since the stdlib does provide `std::env::temp_dir()`, whereas `/var/run/user/[$uid]` is not supported in the standard lib (as far as I can see)

Reply:
There doesn't really have to be a helper function for this... You can just create the file. But fine, happy to leave this potentially for a future PR.